File: src/main/aggregator.js
'use strict'
// @ts-check
/**
 * 聚合：把 commit 与笔记按 (日期, 项目) 分桶，并执行笔记融合分配。
 *  - 项目级笔记 → 对应项目桶
 *  - 通用笔记（project=null）→ 注入当天所有桶的 sharedNotes + 兜底【miscProject】独立段
 */
const { parseDateInput } = require('./utils')

/**
 * @param {object[]} commits
 * @param {object[]} notes
 * @param {string} miscProject
 * @param {string[]} repoOrder - 仓库配置中的项目名顺序（用于排序），miscProject 恒排末尾
 * @returns {object[]} buckets[] —— 已按日期升序、项目顺序排序
 */
function aggregate(commits, notes, miscProject = '日常工作', repos = []) {
  const orderIndex = new Map()
  const aliasToName = new Map()
  const nameToDisplay = new Map()
  repos.forEach((r, i) => {
    if (r.name) {
      orderIndex.set(r.name, i)
      nameToDisplay.set(r.name, r.alias || r.name)
    }
    if (r.alias) aliasToName.set(r.alias, r.name)
  })
  const displayOf = (proj) => nameToDisplay.get(proj) || proj
  const normalizeProject = (p) => aliasToName.get(p) || p

  const dayMap = new Map() // dayStr -> Map(project -> bucket)

  const ensureBucket = (dayStr, project) => {
    if (!dayMap.has(dayStr)) dayMap.set(dayStr, new Map())
    const pm = dayMap.get(dayStr)
    if (!pm.has(project)) {
      pm.set(project, {
        day: parseDateInput(dayStr),
        dayStr,
        project,
        displayName: displayOf(project),
        commits: [],
        notes: [],
        sharedNotes: [],
        isNotesOnly: true,
      })
    }
    return pm.get(project)
  }

  // 1) commit 入桶
  for (const c of commits) {
    const proj = c.project || miscProject
    const b = ensureBucket(c.localDate, proj)
    b.commits.push(c)
    b.isNotesOnly = false
  }

  // 2) 笔记入桶 + 通用笔记融合
  for (const n of notes) {
    if (n.project) {
      // 项目级笔记：归对应项目桶（支持别名映射）；即便该桶无 commit，也形成 notes 桶
      const b = ensureBucket(n.date, normalizeProject(n.project))
      b.notes.push(n)
    } else {
      // 通用笔记：注入当天所有已存在桶的 sharedNotes
      const pm = dayMap.get(n.date)
      if (pm && pm.size) {
        for (const b of pm.values()) b.sharedNotes.push(n)
      }
      // 同时作为【miscProject】独立段落兜底（覆盖纯非代码工作）
      const mb = ensureBucket(n.date, miscProject)
      mb.notes.push(n)
    }
  }

  // 3) 展平 + 排序（日期升序；同日内 miscProject 排末尾，其余按 repoOrder）
  const cmpProject = (a, b) => {
    if (a === miscProject && b !== miscProject) return 1
    if (b === miscProject && a !== miscProject) return -1
    const ia = orderIndex.has(a) ? orderIndex.get(a) : 9999
    const ib = orderIndex.has(b) ? orderIndex.get(b) : 9999
    if (ia !== ib) return ia - ib
    return a.localeCompare(b)
  }

  const buckets = []
  for (const dayStr of [...dayMap.keys()].sort()) {
    const pm = dayMap.get(dayStr)
    for (const project of [...pm.keys()].sort(cmpProject)) {
      buckets.push(pm.get(project))
    }
  }
  return buckets
}

module.exports = { aggregate }

File: src-tauri/src/aggregator.rs
//! 聚合：对齐 src/main/aggregator.js。
//! 把 commit 与笔记按 (日期, 项目) 分桶，并执行笔记融合分配：
//!  - 项目级笔记 → 对应项目桶（支持别名映射）
//!  - 通用笔记（project=null）→ 注入当天所有桶的 sharedNotes + 兜底【miscProject】独立段

use std::cmp::Ordering;
use std::collections::HashMap;

use chrono::NaiveDate;
use serde::Serialize;

use crate::git::Commit;
use crate::notes::Note;
use crate::utils;

/// 聚合桶（对齐 aggregator.js 的 bucket 结构）。
#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Bucket {
    #[serde(skip)]
    pub day: NaiveDate,
    pub day_str: String,
    pub project: String,
    pub display_name: String,
    pub commits: Vec<Commit>,
    pub notes: Vec<Note>,
    pub shared_notes: Vec<Note>,
    pub is_notes_only: bool,
}

/// 仓库展示信息（聚合排序/别名用，取自 cfg.repos）。
pub struct RepoMeta {
    pub name: String,
    pub alias: String,
}

fn epoch() -> NaiveDate {
    NaiveDate::from_ymd_opt(1970, 1, 1).unwrap()
}

/// 确保 (dayStr, project) 桶存在并返回可变引用（对齐 aggregator.js ensureBucket）。
fn ensure_bucket<'a>(
    day_map: &'a mut HashMap<String, HashMap<String, Bucket>>,
    day_str: &str,
    project: &str,
    display: &str,
) -> &'a mut Bucket {
    let pm = day_map.entry(day_str.to_string()).or_default();
    pm.entry(project.to_string()).or_insert_with(|| Bucket {
        day: utils::parse_date(day_str).unwrap_or_else(|_| epoch()),
        day_str: day_str.to_string(),
        project: project.to_string(),
        display_name: display.to_string(),
        commits: Vec::new(),
        notes: Vec::new(),
        shared_notes: Vec::new(),
        is_notes_only: true,
    })
}

/// 聚合 commit + 笔记为桶数组（已按日期升序、项目顺序排序）。
pub fn aggregate(
    commits: Vec<Commit>,
    notes: Vec<Note>,
    misc_project: &str,
    repos: &[RepoMeta],
) -> Vec<Bucket> {
    let mut order_index: HashMap<String, usize> = HashMap::new();
    let mut alias_to_name: HashMap<String, String> = HashMap::new();
    let mut name_to_display: HashMap<String, String> = HashMap::new();
    for (i, r) in repos.iter().enumerate() {
        if !r.name.is_empty() {
            order_index.insert(r.name.clone(), i);
            name_to_display.insert(
                r.name.clone(),
                if r.alias.is_empty() {
                    r.name.clone()
                } else {
                    r.alias.clone()
                },
            );
        }
        if !r.alias.is_empty() {
            alias_to_name.insert(r.alias.clone(), r.name.clone());
        }
    }
    let display_of =
        |proj: &str| name_to_display.get(proj).cloned().unwrap_or_else(|| proj.to_string());
    let normalize_project =
        |p: &str| alias_to_name.get(p).cloned().unwrap_or_else(|| p.to_string());

    let mut day_map: HashMap<String, HashMap<String, Bucket>> = HashMap::new();

    // 1) commit 入桶
    for c in commits {
        let proj = if c.project.is_empty() {
            misc_project.to_string()
        } else {
            c.project.clone()
        };
        let disp = display_of(&proj);
        let b = ensure_bucket(&mut day_map, &c.local_date, &proj, &disp);
        b.commits.push(c);
        b.is_notes_only = false;
    }

    // 2) 笔记入桶 + 通用笔记融合
    for n in notes {
        match n.project.as_deref() {
            Some(p) if !p.is_empty() => {
                // 项目级笔记：归对应项目桶（支持别名映射）
                let np = normalize_project(p);
                let disp = display_of(&np);
                let b = ensure_bucket(&mut day_map, &n.date, &np, &disp);
                b.notes.push(n);
            }
            _ => {
                // 通用笔记：注入当天所有已存在桶的 sharedNotes
                if let Some(pm) = day_map.get_mut(&n.date) {
                    for b in pm.values_mut() {
                        b.shared_notes.push(n.clone());
                    }
                }
                // 同时作为【miscProject】独立段落兜底
                let disp = display_of(misc_project);
                let mb = ensure_bucket(&mut day_map, &n.date, misc_project, &disp);
                mb.notes.push(n);
            }
        }
    }

    // 3) 展平 + 排序（日期升序；同日内 miscProject 排末尾，其余按 repoOrder）
    let cmp_project = |a: &str, b: &str| -> Ordering {
        let a_misc = a == misc_project;
        let b_misc = b == misc_project;
        if a_misc && !b_misc {
            return Ordering::Greater;
        }
        if b_misc && !a_misc {
            return Ordering::Less;
        }
        let ia = order_index.get(a).copied().unwrap_or(9999);
        let ib = order_index.get(b).copied().unwrap_or(9999);
        if ia != ib {
            return ia.cmp(&ib);
        }
        a.cmp(b)
    };

    let mut day_entries: Vec<(String, HashMap<String, Bucket>)> = day_map.into_iter().collect();
    day_entries.sort_by(|a, b| a.0.cmp(&b.0));
    let mut buckets = Vec::new();
    for (_day, pm) in day_entries {
        let mut items: Vec<(String, Bucket)> = pm.into_iter().collect();
        items.sort_by(|a, b| cmp_project(&a.0, &b.0));
        for (_p, b) in items {
            buckets.push(b);
        }
    }
    buckets
}
